gostream/ffmpeg/avcodec.go

```diff
@@ -0,0 +1,679 @@
+		return "MPEG4"
+	case CodecIDRAWVIDEO:
+		return "RAWVIDEO"
+	case CodecIDMSMPEG4V1:
+		return "MSMPEG4V1"
+	case CodecIDMSMPEG4V2:
+		return "MSMPEG4V2"
+	case CodecIDMSMPEG4V3:
+		return "MSMPEG4V3"
+	case CodecIDWMV1:
+		return "WMV1"
+	case CodecIDWMV2:
+		return "WMV2"
+	case CodecIDH263P:
+		return "H263P"
+	case CodecIDH263I:
+		return "H263I"
+	case CodecIDFLV1:
+		return "FLV1"
+	case CodecIDSVQ1:
+		return "SVQ1"
+	case CodecIDSVQ3:
+		return "SVQ3"
+	case CodecIDDVVIDEO:
+		return "DVVIDEO"
+	case CodecIDHUFFYUV:
+		return "HUFFYUV"
+	case CodecIDCYUV:
+		return "CYUV"
+	case CodecIDH264:
+		return "H264"
+	case CodecIDINDEO3:
+		return "INDEO3"
+	case CodecIDVP3:
+		return "VP3"
+	default:
+		return "unknown"
+	}
+}
+
+// FindDecoder Find a registered decoder with a matching codec ID.
+//
+// @param id AVCodecID of the requested decoder
+// @return A decoder if one was found, NULL otherwise.
+func FindDecoder(ctx context.Context, id CodecID) (Codec, error) {
+	var zero Codec
+	select {
+	case <-ctx.Done():
+		return zero, ctx.Err()
+	default:
+		p := C.avcodec_find_decoder((C.enum_AVCodecID)(id))
+		if p == nil {
+			return zero, errors.Errorf("cannot find decoder %s(%d)", id.ToString(), int(id))
+		}
+		return Codec{
+			p:  p,
+			pp: &p,
+		}, nil
+	}
+}
+
+// PacketAlloc Allocate an AVPacket and set its fields to default values.  The resulting
+// struct must be freed using av_packet_free().
+//
+// @return An AVPacket filled with default values or NULL on failure.
+//
+// @note this only allocates the AVPacket itself, not the data buffers. Those
+// must be allocated through other means such as av_new_packet.
+//
+// @see av_new_packet
+func PacketAlloc() (Packet, error) {
+	p := C.av_packet_alloc()
+	if p == nil {
+		return Packet{}, errors.New("cannot allocate packet")
+	}
+
+	return Packet{
+		p:  p,
+		pp: &p,
+	}, nil
+}
+
+// AllocContext3 Allocate an AVCodecContext and set its fields to default values. The
+// resulting struct should be freed with avcodec_free_context().
+//
+// @param codec if non-NULL, allocate private data and initialize defaults
+//
+//	for the given codec. It is illegal to then call avcodec_open2()
+//	with a different codec.
+//	If NULL, then the codec-specific defaults won't be initialized,
+//	which may result in suboptimal default settings (this is
+//	important mainly for encoders, e.g. libx264).
+//
+// @return An AVCodecContext filled with default values or NULL on failure.
+func (c *Codec) AllocContext3(ctx context.Context) (CodecContext, error) {
+	var zero CodecContext
+	select {
+	case <-ctx.Done():
+		return zero, ctx.Err()
+	default:
+		p := C.avcodec_alloc_context3(c.p)
+		if p == nil {
+			return zero, errors.New("cannot allocate context from codec")
+		}
+		return CodecContext{
+			p:  p,
+			pp: &p,
+		}, nil
+	}
+}
+
+// FreeContext Free the codec context and everything associated with it and write NULL to
+// the provided pointer.
+func (ctxt *CodecContext) FreeContext() {
```

Comment:
The documentation is copied verbatim from the header files. I didn't want to deviate from that.